From 941525948399759db81716676f9cb43b7b25a81d Mon Sep 17 00:00:00 2001
From: Xiao Chen <xiao@cloudera.com>
Date: Mon, 3 Sep 2018 22:20:14 -0700
Subject: [PATCH 2834/2863] CDH-71668. HDFS-13611. Unsafe use of Text as a
 ConcurrentHashMap key in PBHelperClient.

(cherry picked from commit c9b63deb533274ca8ef4939f6cd13f728a067f7b)
(cherry picked from commit 097347c857b5a74b228db2b0d076cde2c767672d)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocolPB/PBHelperClient.java

==C5_FEATURE_IMPALA_METADATA==

Change-Id: I9232f67679d11ea16687d0e78014aa366be8f6ce
---
 .../apache/hadoop/hdfs/protocolPB/PBHelper.java    |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/PBHelper.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/PBHelper.java
index fb457b7..c3787af 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/PBHelper.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocolPB/PBHelper.java
@@ -291,7 +291,7 @@ private static ByteString getFixedByteString(Text key) {
     ByteString value = fixedByteStringCache.get(key);
     if (value == null) {
       value = ByteString.copyFromUtf8(key.toString());
-      fixedByteStringCache.put(key, value);
+      fixedByteStringCache.put(new Text(key.copyBytes()), value);
     }
     return value;
   }
-- 
1.7.9.5

